backend/bytecode.proto

```diff
@@ -0,0 +1,206 @@
+    }
+}
+
+// e
+message Expression {
+    message NewExpression {
+        required Type type = 1;
+        required string this = 2;
+        repeated Declaration declarations = 3;
+    }
+    message CallExpression {
+        required Expression callee = 1;
+        required string method = 2;
+        repeated Expression arguments = 3;
+    }
+    message AccessExpression {
+        required Expression expression = 1;
+        required string field = 2;
+    }
+    message AssignmentExpression {
+        required Expression expression = 1;
+        required string field = 2;
+        required Expression value = 3;
+    }
+    message LetExpression {
+        required string variable = 1;
+        required Expression value = 2;
+        required Expression expression = 3;
+    }
+    message Literal {
+        oneof literal {
+            string string = 1;
+            bytes integer = 2;
+        }
+    }
+    message MatchExpression {
+        message MatchArm {
+            required string variable = 1;
+            required string path = 2;
+            required Expression expression = 3;
+        }
+        required Expression expression = 1;
+        repeated MatchArm arms = 2;
+        optional Expression else = 3;
+    }
+    oneof expression {
+        string variable = 1;
+        NewExpression newExpression = 2;
+        CallExpression callExpression = 3;
+        AccessExpression accessExpression = 4;
+        AssignmentExpression assignmentExpression = 5;
+        LetExpression letExpression = 6;
+        Literal literal = 7;
+    }
+}
+
+// d
+message Declaration {
+    message VariableDeclaration {
+        required VariableDeclarationType declarationType = 1;
+        required string variable = 2;
+        required Type type = 3;
+        required string value = 4;
+    }
+    message MethodDeclaration {
+        required string methodName = 1;
+        repeated MethodArgument arguments = 2;
+        required Type returnType = 3;
+        required Expression body = 4;
+    }
+    oneof declaration {
+        VariableDeclaration variableDeclaration = 1;
+        MethodDeclaration methodDeclaration = 2;
+        TypeDeclaration typeDeclaration = 3;
+    }
+}
+
+// T
+message TypeDesc {
+    // c
+    message CaseDesc {
+        optional string extends = 1;
+        required Type type = 2;
+    }
+    message ExTag {
+        // empty?
+    }
+    message DataTag {
+        repeated string paths = 1;
+    }
+    required CaseDesc caseDesc = 1;
+    optional ExTag extag = 2;
+    optional DataTag datatag = 3;
+    optional Expression metaData = 4;
+}
+
+// \tau
+message Type {
+    // \tau { x => \overline{\sigma} }_s
+    message CompoundType { // better name?
+        required Type base = 1; // better name?
+        required string this = 2;
+        repeated DeclType declTypes = 3;
+        required Statefulness statefulness = 4;
+    }
+    enum SpecialType { // better name? PrimitiveType?
```

Comment:
SpecialType is ok.  PrimitiveType makes me think of int/float.  Maybe SimpleType?